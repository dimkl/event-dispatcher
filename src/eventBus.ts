export const globalEventBus = new EventTarget();
