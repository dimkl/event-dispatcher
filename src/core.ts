import { debounce } from "./utils";
import { globalEventBus } from "./eventBus";
import { CustomEvent } from "./customEvent";

type DecoratorType = (
  target: any,
  propertyKey: string,
  descriptor: PropertyDescriptor
) => any;

type DecoratorReturnFnType = (
  target: any,
  propertyKey?: string,
  descriptor?: PropertyDescriptor
) => any;

type BaseParams = {
  eventName?: string;
  eventBus?: EventTarget;
};

type OnOptions = {
  once?: boolean;
  debounceMs?: number;
  errorHandler?: (err: Error) => void;
};

type OnParams = BaseParams & OnOptions;
type DispatchParams = BaseParams;

export const dispatch = ({
  eventName,
  eventBus = globalEventBus,
}: DispatchParams = {}) => {

  const dispatcher: DecoratorType = (target, propertyKey, descriptor) => {
    const event = eventName || generateEventName(target, propertyKey, "");

    function wrapperFn(...args: any[]): typeof descriptor.value {
      // When static method use null and when instance method use this
      // @ts-ignore
      const applyTarget = isInstanceMethod(target) ? this : null;
      const result = descriptor.value.apply(applyTarget, args);
      Promise.resolve(result).then((data) => {
        eventBus.dispatchEvent(new CustomEvent(event, { data }));
      });
      return result;
    };

    Object.defineProperty(target, propertyKey, {
      value: wrapperFn,
      configurable: true,
      writable: true,
    });

    return wrapperFn;
  };

  const dispatchMethod: DecoratorType = (...args) => {
    return {
      get() {
        return dispatcher(...args);
      }
    }
  };

  const dispatchClass = (constructor: any) => {
    decorateFnToAllMethods(constructor, dispatcher);
    return constructor;
  };

  const returnFn: DecoratorReturnFnType = (target, propertyKey, descriptor) => {
    return propertyKey && descriptor
      ? dispatchMethod(target, propertyKey, descriptor)
      : dispatchClass(target);
  };

  return returnFn;
};

export const on = ({
  eventName,
  eventBus = globalEventBus,
  errorHandler,
  once,
  debounceMs,
}: OnParams = {}) => {
  const onClass = (constructor: any) => {
    decorateFnToAllMethods(constructor, onMethod);
    return constructor;
  };

  const onMethod: DecoratorType = (target, propertyKey, descriptor) => {
    const event = eventName || generateEventName(target, propertyKey, "Handler");

    let eventHandler = (evt: unknown) => {
      try {
        // When static method use null and when instance method use this
        const applyTarget = isInstanceMethod(target) ? this : null;
        descriptor.value.apply(applyTarget, [evt]);
      } catch (err) {
        if (errorHandler) {
          errorHandler(err as Error);
        } else {
          eventBus.dispatchEvent(new CustomEvent("error", { data: err }));
        }
      }

      if (once) {
        eventBus.removeEventListener(event, eventHandler);
      }
    };

    if (debounceMs) {
      eventHandler = debounce(eventHandler, debounceMs);
    }
    eventBus.addEventListener(event, eventHandler);
  };

  const returnFn: DecoratorReturnFnType = (target, propertyKey, descriptor) => {
    return propertyKey && descriptor
      ? onMethod(target, propertyKey, descriptor)
      : onClass(target);
  };

  return returnFn;
};

type DecorateHandlerType = (t: any, m: string, p: PropertyDescriptor) => unknown;

const decorateMethod = (target: any, methodName: string, handler: DecorateHandlerType) => {
  const propertyDescriptor = Object.getOwnPropertyDescriptor(target, methodName);
  /* istanbul ignore next */
  if (!propertyDescriptor) return;

  target[methodName] = handler(target, methodName, propertyDescriptor);
}

const decorateFnToAllMethods = (constructor: any, handler: DecorateHandlerType) => {
  getStaticMethods(constructor).forEach(methodName => decorateMethod(constructor, methodName, handler));
  getInstanceMethods(constructor).forEach(methodName => decorateMethod(constructor.prototype, methodName, handler));
};

const isInstanceMethod = (target: any): boolean => {
  return !target.prototype;
}

const getNamespace = (target: any, suffix: string): string => {
  const Fn = isInstanceMethod(target) ? target.constructor : target;
  return Fn.name.replace(suffix, "").toLowerCase();
}

const generateEventName = (target: any, propertyKey: string, suffix: string): string => {
  return [getNamespace(target, suffix), propertyKey].filter(Boolean).join(":");
}

const getStaticMethods = (constructor: Function): string[] => {
  return Object.getOwnPropertyNames(constructor).filter(
    (m) => !["constructor", "length", "name", "prototype"].includes(m)
  );
}

const getInstanceMethods = (constructor: Function): string[] => {
  return Object.getOwnPropertyNames(constructor.prototype).filter((m) => !["constructor"].includes(m));
}