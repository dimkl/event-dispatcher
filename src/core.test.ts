import { on, dispatch, CustomEvent } from "./core";
import { globalEventBus } from "./eventBus";

describe("core", () => {
  describe("CustomEvent", () => {
    test("extends Event", () => {
      expect(new CustomEvent("eventName")).toBeInstanceOf(Event);
    });

    test("holds data", () => {
      const c = new CustomEvent("eventName");
      expect(c.data).toBeUndefined();

      const data = { aloha: 1 };
      const c2 = new CustomEvent("eventName", { data });
      expect(c2.data).toMatchObject(data);
    });
  });

  describe("@on()", () => {
    test("listen events (implicitly)", () => {
      const listenerSpy = jest.fn((_evt: string) => {});
      class Example {
        @on()
        static eventA() {
          listenerSpy("res example:eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res example:eventB");
        }
      }

      class ExampleB {
        @on()
        static eventA() {
          listenerSpy("res exampleb:eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res exampleb:eventB");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));

      expect(listenerSpy.mock.calls).toMatchSnapshot();
    });

    test("listen events (explicitly)", () => {
      const listenerSpy = jest.fn((_evt: string) => {});
      class Example {
        @on({ eventName: "eventA" })
        static eventA() {
          listenerSpy("res eventA");
        }

        @on({ eventName: "eventB" })
        static eventB() {
          listenerSpy("res eventB");
        }
      }

      class ExampleB {
        @on({ eventName: "eventC" })
        static eventA() {
          listenerSpy("res eventC");
        }

        @on({ eventName: "eventD" })
        static eventB() {
          listenerSpy("res eventD");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("eventD"));

      expect(listenerSpy.mock.calls).toMatchSnapshot();
    });

    test("listen event once", () => {
      const listenerSpy = jest.fn((_evt: string) => {});
      class Example {
        @on({ once: true })
        static eventA() {
          listenerSpy("res example:eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res example:eventB");
        }
      }

      class ExampleB {
        @on({ eventName: "eventC", once: true })
        static eventA() {
          listenerSpy("res eventC");
        }

        @on()
        static eventB() {
          listenerSpy("res exampleb:eventB");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));

      expect(listenerSpy.mock.calls).toMatchSnapshot();
    });

    test("listen event debounced", async () => {
      const listenerSpy = jest.fn((_evt: string) => {});
      class Example {
        @on({ debounceMs: 1 })
        static eventA() {
          listenerSpy("res example:eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res example:eventB");
        }
      }

      class ExampleB {
        @on({ eventName: "eventC", debounceMs: 1 })
        static eventA() {
          listenerSpy("res eventC");
        }

        @on()
        static eventB() {
          listenerSpy("res exampleb:eventB");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));

      await new Promise((res) => setTimeout(res, 1));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      await new Promise((res) => setTimeout(res, 1));

      expect(listenerSpy.mock.calls).toMatchSnapshot();
    });

    test.skip("listen event errorHandler", async () => {
      const listenerSpy = jest.fn((_evt: string) => {});
      const errorHandlerSpy = jest.fn(() => {});

      class Example {
        @on({ errorHandler: errorHandlerSpy })
        static eventA() {
          listenerSpy("res example:eventA");
        }

        @on({ errorHandler: errorHandlerSpy })
        static eventB() {
          throw new Error("err example:eventB");
          listenerSpy("res example:eventB");
        }
      }

      class ExampleError {
        @on({ eventName: "eventC", errorHandler: errorHandlerSpy })
        static eventA() {
          throw new Error("err eventC");
          listenerSpy("res eventC");
        }

        @on()
        static eventB() {
          throw new Error("err exampleerror:eventB");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("example:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("eventC"));
      await expect(() =>
        globalEventBus.dispatchEvent(new CustomEvent("exampleerror:eventB"))
      ).toThrow("err exampleerror:eventB");

      expect(listenerSpy.mock.calls).toMatchSnapshot();
      expect(errorHandlerSpy.mock.calls).toMatchSnapshot();
    });

    test("listen event on different eventBus", () => {
      const eventBus = new EventTarget();

      const listenerSpy = jest.fn((_evt: string) => {});
      class Example {
        @on({ eventName: "eventA", eventBus })
        static eventA() {
          listenerSpy("res eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res example:eventB");
        }
      }

      class ExampleB {
        @on({ eventBus })
        static eventA() {
          listenerSpy("res exampleb:eventA");
        }

        @on()
        static eventB() {
          listenerSpy("res exampleb:eventB");
        }
      }

      globalEventBus.dispatchEvent(new CustomEvent("eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("example:eventB"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventA"));
      globalEventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));
      expect(listenerSpy.mock.calls).toMatchSnapshot();
      listenerSpy.mockClear();

      eventBus.dispatchEvent(new CustomEvent("eventA"));
      eventBus.dispatchEvent(new CustomEvent("example:eventB"));
      eventBus.dispatchEvent(new CustomEvent("exampleb:eventA"));
      eventBus.dispatchEvent(new CustomEvent("exampleb:eventB"));

      expect(listenerSpy.mock.calls).toMatchSnapshot();
    });

    test.todo("listen event on debounced once with different eventBus");
  });

  describe("@dispatch()", () => {
    test("dispatch events (implicitly)", async () => {
      class Example {
        @dispatch()
        eventA() {
          return "res example:eventA";
        }

        @dispatch()
        eventB() {
          return "res example:eventB";
        }
      }

      class ExampleB {
        @dispatch()
        eventA() {
          return "res exampleb:eventA";
        }

        @dispatch()
        eventB() {
          return "res exampleb:eventB";
        }
      }

      const dispatcherSpy = jest.fn();
      globalEventBus.addEventListener("example:eventA", dispatcherSpy);
      globalEventBus.addEventListener("example:eventB", dispatcherSpy);
      globalEventBus.addEventListener("exampleb:eventA", dispatcherSpy);
      globalEventBus.addEventListener("exampleb:eventB", dispatcherSpy);

      expect(dispatcherSpy).not.toBeCalled();

      const e = new Example();
      e.eventA();
      e.eventB();
      const eB = new ExampleB();
      eB.eventA();
      eB.eventB();

      // wait for the Promise used in dispatching events
      await new Promise((res) => setTimeout(res, 1));

      expect(dispatcherSpy).toBeCalledTimes(4);
      expect(dispatcherSpy.mock.calls).toMatchObject([
        [expect.objectContaining({ data: "res example:eventA" })],
        [expect.objectContaining({ data: "res example:eventB" })],
        [expect.objectContaining({ data: "res exampleb:eventA" })],
        [expect.objectContaining({ data: "res exampleb:eventB" })],
      ]);
    });

    test("dispatch events (explicitly)", async () => {
      class Example {
        @dispatch({ eventName: "eventA" })
        eventA() {
          return "res eventA";
        }

        @dispatch({ eventName: "eventB" })
        eventB() {
          return "res eventB";
        }
      }

      class ExampleB {
        @dispatch({ eventName: "eventC" })
        eventA() {
          return "res eventC";
        }

        @dispatch({ eventName: "eventD" })
        eventB() {
          return "res eventD";
        }
      }

      const dispatcherSpy = jest.fn();
      globalEventBus.addEventListener("eventA", dispatcherSpy);
      globalEventBus.addEventListener("eventB", dispatcherSpy);
      globalEventBus.addEventListener("eventC", dispatcherSpy);
      globalEventBus.addEventListener("eventD", dispatcherSpy);

      expect(dispatcherSpy).not.toBeCalled();

      const e = new Example();
      e.eventA();
      e.eventB();
      const eB = new ExampleB();
      eB.eventA();
      eB.eventB();

      // wait for the Promise used in dispatching events
      await new Promise((res) => setTimeout(res, 1));

      expect(dispatcherSpy).toBeCalledTimes(4);
      expect(dispatcherSpy.mock.calls).toMatchObject([
        [expect.objectContaining({ data: "res eventA" })],
        [expect.objectContaining({ data: "res eventB" })],
        [expect.objectContaining({ data: "res eventC" })],
        [expect.objectContaining({ data: "res eventD" })],
      ]);
    });

    test("dispatch event on different eventBus", async () => {
      const eventBus = new EventTarget();
      class Example {
        @dispatch({ eventName: "eventA", eventBus })
        eventA() {
          return "res eventA";
        }

        @dispatch({ eventName: "eventB" })
        eventB() {
          return "res eventB";
        }
      }

      class ExampleB {
        @dispatch({ eventBus })
        eventA() {
          return "res exampleb:eventA";
        }

        @dispatch()
        eventB() {
          return "res exampleb:eventB";
        }
      }

      const globalDispatcherSpy = jest.fn();
      globalEventBus.addEventListener("eventA", globalDispatcherSpy);
      globalEventBus.addEventListener("eventB", globalDispatcherSpy);
      globalEventBus.addEventListener("exampleb:eventA", globalDispatcherSpy);
      globalEventBus.addEventListener("exampleb:eventB", globalDispatcherSpy);

      const dispatcherSpy = jest.fn();
      eventBus.addEventListener("eventA", dispatcherSpy);
      eventBus.addEventListener("eventB", dispatcherSpy);
      eventBus.addEventListener("exampleb:eventA", dispatcherSpy);
      eventBus.addEventListener("exampleb:eventB", dispatcherSpy);

      expect(globalDispatcherSpy).not.toBeCalled();
      expect(dispatcherSpy).not.toBeCalled();

      const e = new Example();
      e.eventA();
      e.eventB();
      const eB = new ExampleB();
      eB.eventA();
      eB.eventB();

      // wait for the Promise used in dispatching events
      await new Promise((res) => setTimeout(res, 1));

      expect(globalDispatcherSpy).toBeCalledTimes(2);
      expect(globalDispatcherSpy.mock.calls).toMatchObject([
        [expect.objectContaining({ data: "res eventB" })],
        [expect.objectContaining({ data: "res exampleb:eventB" })],
      ]);

      expect(dispatcherSpy).toBeCalledTimes(2);
      expect(dispatcherSpy.mock.calls).toMatchObject([
        [expect.objectContaining({ data: "res eventA" })],
        [expect.objectContaining({ data: "res exampleb:eventA" })],
      ]);
    });
  });
});
