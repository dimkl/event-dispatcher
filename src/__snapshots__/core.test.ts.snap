// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core @on() listen event debounced 1`] = `
[
  [
    "res example:eventB",
  ],
  [
    "res example:eventB",
  ],
  [
    "res exampleb:eventB",
  ],
  [
    "res exampleb:eventB",
  ],
  [
    "res example:eventA",
  ],
  [
    "res eventC",
  ],
  [
    "res example:eventA",
  ],
  [
    "res eventC",
  ],
]
`;

exports[`core @on() listen event errorHandler 1`] = `
[
  [
    "res example:eventA",
  ],
]
`;

exports[`core @on() listen event errorHandler 2`] = `
[
  [
    [Error: err example:eventB],
  ],
  [
    [Error: err eventC],
  ],
]
`;

exports[`core @on() listen event on different eventBus 1`] = `
[
  [
    "res example:eventB",
  ],
  [
    "res exampleb:eventB",
  ],
]
`;

exports[`core @on() listen event on different eventBus 2`] = `
[
  [
    "res eventA",
  ],
  [
    "res exampleb:eventA",
  ],
]
`;

exports[`core @on() listen event once 1`] = `
[
  [
    "res example:eventA",
  ],
  [
    "res example:eventB",
  ],
  [
    "res example:eventB",
  ],
  [
    "res eventC",
  ],
  [
    "res exampleb:eventB",
  ],
  [
    "res exampleb:eventB",
  ],
]
`;

exports[`core @on() listen events (explicitly) 1`] = `
[
  [
    "res eventA",
  ],
  [
    "res eventB",
  ],
  [
    "res eventC",
  ],
  [
    "res eventD",
  ],
]
`;

exports[`core @on() listen events (implicitly) 1`] = `
[
  [
    "res example:eventA",
  ],
  [
    "res example:eventB",
  ],
  [
    "res exampleb:eventA",
  ],
  [
    "res exampleb:eventB",
  ],
]
`;
